core-rust/state-manager/src/state_manager.rs

```diff
@@ -143,7 +143,7 @@ impl StateManager {
         mempool_relay_dispatcher: Option<MempoolRelayDispatcher>,
         lock_factory: &LockFactory,
         metrics_registry: &Registry,
-        scheduler: &impl Scheduler,
+        scheduler: &Scheduler<impl Spawner, impl Tracker, impl Metrics>,
```

Comment:
_(note from author)_

This is the least “convenient” spot of the scheduler facade that I ended up with.

Basically, I’d prefer to have just `&Scheduler` here - even `&Scheduler<_, _, _>` (if it compiled) would be sub-optimal (since I would have to update this line when I add 1 more capability to the facade).

(e.g. I considered putting a trait over the entire `Scheduler` struct. But its “builder” nature does not make it so easy, and AFAIS we would lose some type information [i.e. which capabilities are configured] also in the places where we _do_ care about them. I basically want a concise way of saying "give me a `Scheduler` and I don't care what type params it has").

Any suggestions are welcome, but worst-case scenario I can live with what we have here.

Reply:
Yeah, it looks a bit ugly. Unfortunately type aliases with `impl` are unstable (`type GenericScheduler = Scheduler<impl Spawner, impl Tracker, impl Metrics>`) - let's hope it doesn't take years to get in.

Because in `StateManager` you don't really care about any other capability than `Spawner` you could pass "something" else that only has that - but you lose the whole hierarchical (builder) nature and `named(..)`.

I actually like this pattern and how it ended up, even with this inconvenience. I am fine to go forward hoping that https://github.com/rust-lang/rust/issues/63063 will be sorted out.